src/doc/octree_map.h

```diff
@@ -0,0 +1,182 @@
+// Aseprite
+// Copyright (c) 2019 Igara Studio S.A.
+//
+// This file is released under the terms of the MIT license.
+// Read LICENSE.txt for more information.
+
+#ifndef DOC_OCTREEMAP_H_INCLUDED
+#define DOC_OCTREEMAP_H_INCLUDED
+#pragma once
+
+#include "doc/image_impl.h"
+#include "doc/palette.h"
+
+#include <vector>
+#include <array>
+
+namespace doc {
+
+class OctreeMap;
+class OctreeNode;
+
+class OctreeNode {
+private:
+  class LeafColor {
+  public:
+    LeafColor() :
+      m_r(0),
+      m_g(0),
+      m_b(0),
+      m_pixelCount(0)
+    {}
+
+    LeafColor(int r, int g, int b, int pixelCount) :
+      m_r((double)r),
+      m_g((double)g),
+      m_b((double)b),
+      m_pixelCount(pixelCount)
+    {}
+
+    ~LeafColor() {}
+
+    void add(color_t c)
+    {
+      m_r += rgba_getr(c);
+      m_g += rgba_getg(c);
+      m_b += rgba_getb(c);
+      m_pixelCount++;
+    }
+
+    void add(LeafColor leafColor)
+    {
+      m_r += leafColor.m_r;
+      m_g += leafColor.m_g;
+      m_b += leafColor.m_b;
+      m_pixelCount += leafColor.m_pixelCount;
+    }
+
+    LeafColor normalizeColor()
+    {
+      int auxR = (((int)m_r) % m_pixelCount > m_pixelCount / 2)? 1 : 0;
+      int auxG = (((int)m_g) % m_pixelCount > m_pixelCount / 2)? 1 : 0;
+      int auxB = (((int)m_b) % m_pixelCount > m_pixelCount / 2)? 1 : 0;
+      return LeafColor(m_r / m_pixelCount + auxR,
+                           m_g / m_pixelCount + auxG,
+                           m_b / m_pixelCount + auxB,
+                           m_pixelCount);
+    }
+
+    color_t LeafColorToColor()
+    {
+      return 0xff000000 + (((int)m_b) << 16) + (((int)m_g) << 8) + (int)m_r;
+    }
+
+    int pixelCount() { return m_pixelCount; }
+
+    private:
+      double m_r;
+      double m_g;
+      double m_b;
+      int m_pixelCount;
+  };
+
+public:
+  OctreeNode()
+  {
+    m_paletteIndex = 0;
+    m_children.reset(nullptr);
+    m_parent = nullptr;
+  };
+
+  ~OctreeNode()
+  {
+  }
+
+  static int getOctet(color_t c, int level)
+  {
+    int aux = c >> (7 - level);
+    int octet = aux & 1;
+    aux = aux >> (7);
+    octet += (aux & 2);
+    return octet + ((aux >> 7) & 4);
```

Comment:
It looks like the `octet` is built bit by bit (1, 2, and then 4), so I guess it can be:
```cpp
    int octet = aux & 1;
    aux >>= 7;   // REMOVETHISCOMMENT check that we are using >>= instead of aux = aux >>
    octet |= (aux & 2);
    return octet | ((aux >> 7) & 4);
```
Not sure but I think this code should be equivalent.